pkg/karmadactl/unjoin.go

```diff
@@ -169,67 +185,98 @@ func UnJoinCluster(controlPlaneRestConfig, clusterConfig *rest.Config, opts Comm
+			clusterRoleName := names.GenerateRoleName(serviceAccountName)
+			clusterRoleBindingName := clusterRoleName
 
-		// delete service account from unjoining cluster
-		err = deleteServiceAccount(clusterKubeClient, opts.ClusterNamespace, opts.ClusterName, opts.forceDeletion, opts.DryRun)
-		if err != nil {
-			klog.Errorf("Failed to delete service account in unjoining cluster %q: %v", opts.ClusterName, err)
-			return err
+			err = deleteRBACResources(clusterKubeClient, clusterRoleName, clusterRoleBindingName, opts.ClusterName, opts.forceDeletion, opts.DryRun)
+			if err != nil {
+				klog.Errorf("Failed to delete RBAC resource in unjoining cluster %q: %v", opts.ClusterName, err)
+				return err
+			}
+
+			// delete service account from unjoining cluster
+			err = deleteServiceAccount(clusterKubeClient, opts.ClusterNamespace, serviceAccountName, opts.ClusterName, opts.forceDeletion, opts.DryRun)
+			if err != nil {
+				klog.Errorf("Failed to delete service account in unjoining cluster %q: %v", opts.ClusterName, err)
+				return err
+			}
+
+			// delete namespace from unjoining cluster
+			err = deleteNamespaceFromUnjoinCluster(clusterKubeClient, opts.ClusterNamespace, opts.ClusterName, opts.forceDeletion, opts.DryRun)
+			if err != nil {
+				klog.Errorf("Failed to delete namespace in unjoining cluster %q: %v", opts.ClusterName, err)
+				return err
+			}
 		}
 
-		// delete namespace from unjoining cluster
-		err = deleteNamespaceFromUnjoinCluster(clusterKubeClient, opts.ClusterNamespace, opts.ClusterName, opts.forceDeletion, opts.DryRun)
-		if err != nil {
-			klog.Errorf("Failed to delete namespace in unjoining cluster %q: %v", opts.ClusterName, err)
-			return err
+		if syncMode == clusterv1alpha1.Pull {
+			// delete RBAC resource from unjoining cluster
+			err = deleteRBACResources(clusterKubeClient, agentClusterRole, agentClusterRoleBinding, opts.ClusterName, opts.forceDeletion, opts.DryRun)
+			if err != nil {
+				klog.Errorf("Failed to delete RBAC resource in unjoining cluster %q: %v", opts.ClusterName, err)
+				return err
+			}
+
+			// delete karmada-agent deployment from unjoining cluster
+			err = deleteAgentDeployment(clusterKubeClient, agentNamespace, agentDeployment, opts.ClusterName, opts.forceDeletion, opts.DryRun)
+			if err != nil {
+				klog.Errorf("Failed to delete service account in unjoining cluster %q: %v", opts.ClusterName, err)
+				return err
+			}
+
+			// delete service account from unjoining cluster
+			err = deleteServiceAccount(clusterKubeClient, agentNamespace, agentServiceAccount, opts.ClusterName, opts.forceDeletion, opts.DryRun)
+			if err != nil {
+				klog.Errorf("Failed to delete service account in unjoining cluster %q: %v", opts.ClusterName, err)
+				return err
+			}
+
+			// delete namespace from unjoining cluster
+			err = deleteNamespaceFromUnjoinCluster(clusterKubeClient, agentNamespace, opts.ClusterName, opts.forceDeletion, opts.DryRun)
```

Comment:
1. I believe deleting namespace would delete all resources under that namespace automatically, so it seems unnecessary to delete karmada-agent sa and deploy, WDYT?
2. There might be an unusual case that the unjoined cluster is `karmada-host`(if user joined `karmada-host` as a member in pull mode), in which case if we delete `karmada-system` namespace, it would break karmada control plane

Reply:
Thanks for you reviews, @dddddai 

> I believe deleting namespace would delete all resources under that namespace automatically, so it seems unnecessary to delete karmada-agent sa and deploy, WDYT?

That is right, but I care about that the namespace isn't deleted correctly, because I've been in this situation a lot when deleting namespace directly, the namespace is still the status `terminating` that is the `Finalizers` problem. So I decide to delete them one by one.

> There might be an unusual case that the unjoined cluster is karmada-host(if user joined karmada-host as a member in pull mode), in which case if we delete karmada-system namespace, it would break karmada control plane

That is also my question. Why we set different namespace in Push mode `karmada-cluster` and Pull mode `karmada-system`? @RainbowMango